File: java/design_pattern/state/MainState.java
/**
 * Comprehensive demonstration of the State Pattern in Java
 *
 * The State pattern allows an object to alter its behavior when its internal
 * state changes. The object will appear to change its class.
 */

// State interface
interface VendingMachineState {
    void insertCoin();
    void ejectCoin();
    void selectProduct();
    void dispense();
}

// Context
class VendingMachine {
    private VendingMachineState noCoinState;
    private VendingMachineState hasCoinState;
    private VendingMachineState soldState;
    private VendingMachineState soldOutState;

    private VendingMachineState currentState;
    private int count;

    public VendingMachine(int count) {
        noCoinState = new NoCoinState(this);
        hasCoinState = new HasCoinState(this);
        soldState = new SoldState(this);
        soldOutState = new SoldOutState(this);

        this.count = count;
        if (count > 0) {
            currentState = noCoinState;
        } else {
            currentState = soldOutState;
        }
    }

    public void insertCoin() { currentState.insertCoin(); }
    public void ejectCoin() { currentState.ejectCoin(); }
    public void selectProduct() {
        currentState.selectProduct();
        currentState.dispense();
    }

    public void setState(VendingMachineState state) { currentState = state; }
    public VendingMachineState getNoCoinState() { return noCoinState; }
    public VendingMachineState getHasCoinState() { return hasCoinState; }
    public VendingMachineState getSoldState() { return soldState; }
    public VendingMachineState getSoldOutState() { return soldOutState; }

    public void releaseProduct() {
        if (count > 0) {
            count--;
            System.out.println("  [Machine] Product dispensed. Remaining: " + count);
        }
    }

    public int getCount() { return count; }
}

// Concrete States
class NoCoinState implements VendingMachineState {
    private VendingMachine machine;

    public NoCoinState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertCoin() {
        System.out.println("  [NoCoin] Coin inserted");
        machine.setState(machine.getHasCoinState());
    }

    @Override
    public void ejectCoin() {
        System.out.println("  [NoCoin] No coin to eject");
    }

    @Override
    public void selectProduct() {
        System.out.println("  [NoCoin] Please insert a coin first");
    }

    @Override
    public void dispense() {
        System.out.println("  [NoCoin] Please insert a coin first");
    }
}

class HasCoinState implements VendingMachineState {
    private VendingMachine machine;

    public HasCoinState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertCoin() {
        System.out.println("  [HasCoin] Coin already inserted");
    }

    @Override
    public void ejectCoin() {
        System.out.println("  [HasCoin] Coin ejected");
        machine.setState(machine.getNoCoinState());
    }

    @Override
    public void selectProduct() {
        System.out.println("  [HasCoin] Product selected");
        machine.setState(machine.getSoldState());
    }

    @Override
    public void dispense() {
        System.out.println("  [HasCoin] Please select a product");
    }
}

class SoldState implements VendingMachineState {
    private VendingMachine machine;

    public SoldState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertCoin() {
        System.out.println("  [Sold] Please wait, dispensing product");
    }

    @Override
    public void ejectCoin() {
        System.out.println("  [Sold] Cannot eject, product is being dispensed");
    }

    @Override
    public void selectProduct() {
        System.out.println("  [Sold] Already dispensing");
    }

    @Override
    public void dispense() {
        machine.releaseProduct();
        if (machine.getCount() > 0) {
            machine.setState(machine.getNoCoinState());
        } else {
            System.out.println("  [Sold] Out of products");
            machine.setState(machine.getSoldOutState());
        }
    }
}

class SoldOutState implements VendingMachineState {
    private VendingMachine machine;

    public SoldOutState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertCoin() {
        System.out.println("  [SoldOut] Machine is sold out, returning coin");
    }

    @Override
    public void ejectCoin() {
        System.out.println("  [SoldOut] No coin to eject");
    }

    @Override
    public void selectProduct() {
        System.out.println("  [SoldOut] Machine is sold out");
    }

    @Override
    public void dispense() {
        System.out.println("  [SoldOut] No product to dispense");
    }
}

// Document workflow example

interface DocumentState {
    void edit(DocumentContext doc);
    void review(DocumentContext doc);
    void approve(DocumentContext doc);
    void reject(DocumentContext doc);
    void publish(DocumentContext doc);
}

class DocumentContext {
    private DocumentState draftState = new DraftState();
    private DocumentState reviewState = new ReviewState();
    private DocumentState approvedState = new ApprovedState();
    private DocumentState publishedState = new PublishedState();

    private DocumentState currentState;
    private String name;

    public DocumentContext(String name) {
        this.name = name;
        this.currentState = draftState;
    }

    public void edit() { currentState.edit(this); }
    public void review() { currentState.review(this); }
    public void approve() { currentState.approve(this); }
    public void reject() { currentState.reject(this); }
    public void publish() { currentState.publish(this); }

    public void setState(String state) {
        switch (state) {
            case "draft": currentState = draftState; break;
            case "review": currentState = reviewState; break;
            case "approved": currentState = approvedState; break;
            case "published": currentState = publishedState; break;
        }
    }

    public String getName() { return name; }
}

class DraftState implements DocumentState {
    @Override
    public void edit(DocumentContext doc) {
        System.out.println("  [Draft] Editing document: " + doc.getName());
    }

    @Override
    public void review(DocumentContext doc) {
        System.out.println("  [Draft] Submitting for review: " + doc.getName());
        doc.setState("review");
    }

    @Override
    public void approve(DocumentContext doc) {
        System.out.println("  [Draft] Cannot approve from draft state");
    }

    @Override
    public void reject(DocumentContext doc) {
        System.out.println("  [Draft] Cannot reject from draft state");
    }

    @Override
    public void publish(DocumentContext doc) {
        System.out.println("  [Draft] Cannot publish from draft state");
    }
}

class ReviewState implements DocumentState {
    @Override
    public void edit(DocumentContext doc) {
        System.out.println("  [Review] Cannot edit during review");
    }

    @Override
    public void review(DocumentContext doc) {
        System.out.println("  [Review] Already in review");
    }

    @Override
    public void approve(DocumentContext doc) {
        System.out.println("  [Review] Document approved: " + doc.getName());
        doc.setState("approved");
    }

    @Override
    public void reject(DocumentContext doc) {
        System.out.println("  [Review] Document rejected, returning to draft");
        doc.setState("draft");
    }

    @Override
    public void publish(DocumentContext doc) {
        System.out.println("  [Review] Cannot publish without approval");
    }
}

class ApprovedState implements DocumentState {
    @Override
    public void edit(DocumentContext doc) {
        System.out.println("  [Approved] Editing, returning to draft");
        doc.setState("draft");
    }

    @Override
    public void review(DocumentContext doc) {
        System.out.println("  [Approved] Already approved");
    }

    @Override
    public void approve(DocumentContext doc) {
        System.out.println("  [Approved] Already approved");
    }

    @Override
    public void reject(DocumentContext doc) {
        System.out.println("  [Approved] Rejecting, returning to draft");
        doc.setState("draft");
    }

    @Override
    public void publish(DocumentContext doc) {
        System.out.println("  [Approved] Publishing document: " + doc.getName());
        doc.setState("published");
    }
}

class PublishedState implements DocumentState {
    @Override
    public void edit(DocumentContext doc) {
        System.out.println("  [Published] Cannot edit published document");
    }

    @Override
    public void review(DocumentContext doc) {
        System.out.println("  [Published] Cannot review published document");
    }

    @Override
    public void approve(DocumentContext doc) {
        System.out.println("  [Published] Already published");
    }

    @Override
    public void reject(DocumentContext doc) {
        System.out.println("  [Published] Cannot reject published document");
    }

    @Override
    public void publish(DocumentContext doc) {
        System.out.println("  [Published] Already published");
    }
}

public class MainState {
    public static void main(String[] args) {
        System.out.println("=== State Pattern Demonstration ===\n");

        // Vending machine example
        System.out.println("--- 1. Vending Machine ---");

        VendingMachine machine = new VendingMachine(2);

        System.out.println("\nAttempt without coin:");
        machine.selectProduct();

        System.out.println("\nNormal purchase:");
        machine.insertCoin();
        machine.selectProduct();

        System.out.println("\nEject coin:");
        machine.insertCoin();
        machine.ejectCoin();

        System.out.println("\nLast product:");
        machine.insertCoin();
        machine.selectProduct();

        System.out.println("\nSold out:");
        machine.insertCoin();
        System.out.println();

        // Document workflow
        System.out.println("--- 2. Document Workflow ---");

        DocumentContext doc = new DocumentContext("Annual Report");

        System.out.println("\nDraft phase:");
        doc.edit();
        doc.publish();  // Cannot publish from draft

        System.out.println("\nSubmit for review:");
        doc.review();
        doc.edit();     // Cannot edit during review

        System.out.println("\nReject and revise:");
        doc.reject();
        doc.edit();
        doc.review();

        System.out.println("\nApprove and publish:");
        doc.approve();
        doc.publish();

        System.out.println("\nTry to modify published:");
        doc.edit();

        System.out.println("\n=== Summary ===");
        System.out.println("State pattern benefits:");
        System.out.println("  - Localizes state-specific behavior");
        System.out.println("  - Makes state transitions explicit");
        System.out.println("  - State objects can be shared");
        System.out.println("\nVs Strategy pattern:");
        System.out.println("  - State: behavior changes based on internal state");
        System.out.println("  - Strategy: client chooses algorithm");
    }
}

File: python/design_pattern/state/main_state.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
State Pattern

The State pattern allows an object to alter its behavior when its internal
state changes. The object will appear to change its class.

Key Components:
- Context: Maintains instance of ConcreteState subclass for current state
- State: Interface for encapsulating state-specific behavior
- ConcreteState: Implements behavior associated with a state
"""

from abc import ABC, abstractmethod
from datetime import datetime


# State interface
class VendingMachineState(ABC):
    """Abstract state for vending machine."""

    @abstractmethod
    def insert_coin(self, machine: "VendingMachine", amount: float) -> str:
        pass

    @abstractmethod
    def select_product(self, machine: "VendingMachine", product: str) -> str:
        pass

    @abstractmethod
    def dispense(self, machine: "VendingMachine") -> str:
        pass

    @abstractmethod
    def cancel(self, machine: "VendingMachine") -> str:
        pass

    @abstractmethod
    def get_state_name(self) -> str:
        pass


# Concrete States
class IdleState(VendingMachineState):
    """Machine is idle, waiting for coins."""

    def insert_coin(self, machine: "VendingMachine", amount: float) -> str:
        machine.add_credit(amount)
        machine.set_state(HasMoneyState())
        return f"Inserted ${amount:.2f}. Credit: ${machine.credit:.2f}"

    def select_product(self, machine: "VendingMachine", product: str) -> str:
        return "Please insert coins first"

    def dispense(self, machine: "VendingMachine") -> str:
        return "Please insert coins and select a product"

    def cancel(self, machine: "VendingMachine") -> str:
        return "No transaction to cancel"

    def get_state_name(self) -> str:
        return "Idle"


class HasMoneyState(VendingMachineState):
    """Machine has money, waiting for product selection."""

    def insert_coin(self, machine: "VendingMachine", amount: float) -> str:
        machine.add_credit(amount)
        return f"Added ${amount:.2f}. Total credit: ${machine.credit:.2f}"

    def select_product(self, machine: "VendingMachine", product: str) -> str:
        if product not in machine.inventory:
            return f"Product '{product}' not available"

        price = machine.get_price(product)
        if machine.credit < price:
            return f"Insufficient credit. {product} costs ${price:.2f}, you have ${machine.credit:.2f}"

        if machine.get_stock(product) <= 0:
            return f"{product} is out of stock"

        machine.selected_product = product
        machine.set_state(DispensingState())
        return f"Selected {product} (${price:.2f}). Dispensing..."

    def dispense(self, machine: "VendingMachine") -> str:
        return "Please select a product first"

    def cancel(self, machine: "VendingMachine") -> str:
        refund = machine.credit
        machine.reset_credit()
        machine.set_state(IdleState())
        return f"Transaction cancelled. Refund: ${refund:.2f}"

    def get_state_name(self) -> str:
        return "HasMoney"


class DispensingState(VendingMachineState):
    """Machine is dispensing selected product."""

    def insert_coin(self, machine: "VendingMachine", amount: float) -> str:
        return "Please wait, dispensing in progress"

    def select_product(self, machine: "VendingMachine", product: str) -> str:
        return "Please wait, dispensing in progress"

    def dispense(self, machine: "VendingMachine") -> str:
        product = machine.selected_product
        price = machine.get_price(product)

        # Dispense product
        machine.reduce_stock(product)
        machine.deduct_credit(price)

        # Calculate change
        change = machine.credit
        message = f"Dispensed: {product}"

        if change > 0:
            message += f"\nChange: ${change:.2f}"
            machine.reset_credit()

        # Transition to appropriate state
        if machine.credit > 0:
            machine.set_state(HasMoneyState())
        else:
            machine.set_state(IdleState())

        machine.selected_product = ""
        return message

    def cancel(self, machine: "VendingMachine") -> str:
        return "Cannot cancel during dispensing"

    def get_state_name(self) -> str:
        return "Dispensing"


class OutOfStockState(VendingMachineState):
    """Machine is out of all products."""

    def insert_coin(self, machine: "VendingMachine", amount: float) -> str:
        return f"Machine is out of stock. Returning ${amount:.2f}"

    def select_product(self, machine: "VendingMachine", product: str) -> str:
        return "Machine is out of stock"

    def dispense(self, machine: "VendingMachine") -> str:
        return "Machine is out of stock"

    def cancel(self, machine: "VendingMachine") -> str:
        return "No transaction to cancel"

    def get_state_name(self) -> str:
        return "OutOfStock"


# Context
class VendingMachine:
    """Context: Vending machine that changes behavior based on state."""

    def __init__(self):
        self._state: VendingMachineState = IdleState()
        self._credit = 0.0
        self._inventory: dict[str, dict] = {
            "Cola": {"price": 1.50, "stock": 5},
            "Chips": {"price": 1.25, "stock": 3},
            "Candy": {"price": 0.75, "stock": 10},
            "Water": {"price": 1.00, "stock": 8},
        }
        self.selected_product = ""

    @property
    def credit(self) -> float:
        return self._credit

    @property
    def inventory(self) -> dict[str, dict]:
        return self._inventory

    def set_state(self, state: VendingMachineState) -> None:
        old_state = self._state.get_state_name()
        self._state = state
        print(f"  [State] {old_state} -> {state.get_state_name()}")

    def get_state_name(self) -> str:
        return self._state.get_state_name()

    def add_credit(self, amount: float) -> None:
        self._credit += amount

    def deduct_credit(self, amount: float) -> None:
        self._credit -= amount

    def reset_credit(self) -> None:
        self._credit = 0.0

    def get_price(self, product: str) -> float:
        return self._inventory[product]["price"]

    def get_stock(self, product: str) -> int:
        return self._inventory[product]["stock"]

    def reduce_stock(self, product: str) -> None:
        self._inventory[product]["stock"] -= 1
        # Check if machine is out of stock
        total_stock = sum(p["stock"] for p in self._inventory.values())
        if total_stock == 0:
            self.set_state(OutOfStockState())

    # Delegated methods
    def insert_coin(self, amount: float) -> str:
        return self._state.insert_coin(self, amount)

    def select_product(self, product: str) -> str:
        return self._state.select_product(self, product)

    def dispense(self) -> str:
        return self._state.dispense(self)

    def cancel(self) -> str:
        return self._state.cancel(self)

    def show_products(self) -> None:
        print("\nAvailable products:")
        for name, info in self._inventory.items():
            status = f"Stock: {info['stock']}" if info["stock"] > 0 else "OUT OF STOCK"
            print(f"  {name}: ${info['price']:.2f} ({status})")


# Document example
class DocumentState(ABC):
    """State for document workflow."""

    @abstractmethod
    def edit(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def review(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def approve(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def reject(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def publish(self, doc: "Document") -> str:
        pass

    @abstractmethod
    def get_state_name(self) -> str:
        pass


class DraftState(DocumentState):
    def edit(self, doc: "Document") -> str:
        doc.modified_at = datetime.now()
        return "Document edited"

    def review(self, doc: "Document") -> str:
        doc.set_state(ReviewState())
        return "Document submitted for review"

    def approve(self, doc: "Document") -> str:
        return "Cannot approve draft. Submit for review first"

    def reject(self, doc: "Document") -> str:
        return "Cannot reject draft"

    def publish(self, doc: "Document") -> str:
        return "Cannot publish draft. Must be approved first"

    def get_state_name(self) -> str:
        return "Draft"


class ReviewState(DocumentState):
    def edit(self, doc: "Document") -> str:
        return "Cannot edit during review"

    def review(self, doc: "Document") -> str:
        return "Already in review"

    def approve(self, doc: "Document") -> str:
        doc.set_state(ApprovedState())
        return "Document approved"

    def reject(self, doc: "Document") -> str:
        doc.set_state(DraftState())
        return "Document rejected. Returned to draft"

    def publish(self, doc: "Document") -> str:
        return "Cannot publish. Must be approved first"

    def get_state_name(self) -> str:
        return "Review"


class ApprovedState(DocumentState):
    def edit(self, doc: "Document") -> str:
        doc.set_state(DraftState())
        return "Document moved back to draft for editing"

    def review(self, doc: "Document") -> str:
        return "Already approved"

    def approve(self, doc: "Document") -> str:
        return "Already approved"

    def reject(self, doc: "Document") -> str:
        doc.set_state(DraftState())
        return "Approval revoked. Returned to draft"

    def publish(self, doc: "Document") -> str:
        doc.set_state(PublishedState())
        doc.published_at = datetime.now()
        return "Document published!"

    def get_state_name(self) -> str:
        return "Approved"


class PublishedState(DocumentState):
    def edit(self, doc: "Document") -> str:
        return "Cannot edit published document. Create new version"

    def review(self, doc: "Document") -> str:
        return "Cannot review published document"

    def approve(self, doc: "Document") -> str:
        return "Cannot approve published document"

    def reject(self, doc: "Document") -> str:
        return "Cannot reject published document"

    def publish(self, doc: "Document") -> str:
        return "Already published"

    def get_state_name(self) -> str:
        return "Published"


class Document:
    """Context for document workflow."""

    def __init__(self, title: str):
        self.title = title
        self._state: DocumentState = DraftState()
        self.created_at = datetime.now()
        self.modified_at = datetime.now()
        self.published_at: datetime | None = None

    def set_state(self, state: DocumentState) -> None:
        self._state = state

    def get_state_name(self) -> str:
        return self._state.get_state_name()

    def edit(self) -> str:
        return self._state.edit(self)

    def review(self) -> str:
        return self._state.review(self)

    def approve(self) -> str:
        return self._state.approve(self)

    def reject(self) -> str:
        return self._state.reject(self)

    def publish(self) -> str:
        return self._state.publish(self)


def main() -> None:
    print("=" * 60)
    print("State Pattern - Vending Machine Demo")
    print("=" * 60)

    # Demo 1: Normal vending machine operation
    print("\n--- Normal Operation ---")
    machine = VendingMachine()
    machine.show_products()

    print("\n--- Transaction 1 ---")
    print(machine.insert_coin(1.00))
    print(machine.insert_coin(0.50))
    print(machine.select_product("Cola"))
    print(machine.dispense())

    print("\n--- Transaction 2 (Insufficient funds) ---")
    print(machine.insert_coin(0.50))
    print(machine.select_product("Chips"))

    print("\n--- Transaction 2 (Add more money) ---")
    print(machine.insert_coin(1.00))
    print(machine.select_product("Chips"))
    print(machine.dispense())

    print("\n--- Transaction 3 (Cancel) ---")
    print(machine.insert_coin(2.00))
    print(machine.cancel())

    print("\n--- Invalid Operations ---")
    print(machine.select_product("Cola"))  # No money
    print(machine.dispense())  # Nothing selected

    # Demo 2: Document workflow
    print("\n" + "=" * 60)
    print("State Pattern - Document Workflow Demo")
    print("=" * 60)

    doc = Document("Design Proposal")

    print("\n--- Document Lifecycle ---")
    print(f"Initial state: {doc.get_state_name()}")

    print(f"\n{doc.edit()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.review()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.edit()}")  # Cannot edit during review

    print(f"\n{doc.reject()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.edit()}")
    print(f"{doc.review()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.approve()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.publish()}")
    print(f"State: {doc.get_state_name()}")

    print(f"\n{doc.edit()}")  # Cannot edit published

    print("\n" + "=" * 60)
    print("Benefits of State Pattern:")
    print("=" * 60)
    print("1. Localizes state-specific behavior")
    print("2. Makes state transitions explicit")
    print("3. State objects can be shared")
    print("4. Eliminates large conditional statements")
    print("5. Follows Single Responsibility Principle")


if __name__ == "__main__":
    main()
